crates/pathfinder/src/rpc/api.rs

```diff
@@ -331,6 +333,53 @@ impl RpcApi {
             .and_then(|x| x)
     }
 
+    /// Fetches [GlobalRoot] of a block by block hash.
+    ///
+    /// Tries storage first, then the sequencer.
+    async fn get_root_by_hash(&self, block_hash: StarknetBlockHash) -> RpcResult<GlobalRoot> {
+        let storage = self.storage.clone();
+
+        let handle = tokio::task::spawn_blocking(move || {
+            let mut connection = storage
+                .connection()
+                .context("Opening database connection")
+                .map_err(internal_server_error)?;
+
+            let transaction = connection
+                .transaction()
+                .context("Creating database transaction")
+                .map_err(internal_server_error)?;
+
+            let root = StarknetBlocksTable::get_root(
+                &transaction,
+                StarknetBlocksBlockId::Hash(block_hash),
+            )
+            .context("Read block from database")
+            .map_err(internal_server_error)?;
+
+            Ok(root)
+        });
+
+        let root = handle
+            .await
+            .context("Database read panic or shutting down")
+            .map_err(internal_server_error)
+            // flatten is unstable
+            .and_then(|x| x)?;
+
+        match root {
+            Some(root) => Ok(root),
+            None => Ok(self
+                .sequencer
+                .block_by_hash(BlockHashOrTag::Hash(block_hash))
+                .await
+                .context("Fetch block from sequencer")
```

Comment:
```suggestion
                .context("Fetch pending - 1 block from sequencer")
```

Or something, just different from the usual.

Reply:
Yep, this is a different situation from the other instances of `.context("Fetch block from sequencer")`.